src/commands/multi_block/monitor.rs

```diff
@@ -273,37 +478,41 @@ where
     .await
     {
         (Ok(sol), dur) => {
-            log::trace!(target: LOG_TARGET, "Mining solution took {}ms", dur.as_millis());
+            log::info!(target: LOG_TARGET, "Mining solution took {}ms for block #{}", dur.as_millis(), block_number);
             prometheus::observe_mined_solution_duration(dur.as_millis() as f64);
             sol
         }
-        (Err(e), _) => {
+        (Err(e), dur) => {
+            log::error!(target: LOG_TARGET, "Mining failed after {}ms: {:?}", dur.as_millis(), e);
             return Err(e);
         }
     };
 
+    // Get latest storage state (chain may have progressed while we were mining)
     let storage_head = utils::storage_at_head(&client, listen).await?;
 
+    // Handle existing submissions
     match get_submission(&storage_head, round, signer.account_id(), n_pages).await? {
         CurrentSubmission::Done(score) => {
-            // We have already submitted the solution with a better score or equal score
+            // We have already submitted the solution with a score
             if !score_passes_strategy(paged_raw_solution.score, score, submission_strategy) {
+                log::debug!(target: LOG_TARGET, "Our new score doesn't beat existing submission, skipping");
                 return Ok(());
             }
-            // Revert previous submission and submit the new one.
+            log::debug!(target: LOG_TARGET, "Reverting previous submission to submit better solution");
             dynamic::bail(&client, &signer, listen).await?;
         }
         CurrentSubmission::Incomplete(s) => {
-            // Submit the missing pages.
             if s.score() == paged_raw_solution.score {
+                // Same score, just submit missing pages
                 let mut missing_pages = Vec::new();
```

Comment:
```suggestion
                let missing_pages = s.get_missing_pages().map(|x| x as usize).map(|p| paged_raw_solution[p].clone()).collect::Vec<_>();
```

just an alternative syntax

Reply:
yeah, I like the functional style more and we also eliminate mutability. Thx, done.